status/discover.go

```diff
@@ -0,0 +1,62 @@
+package status
+
+import (
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/kubernetes"
+	"io/ioutil"
+)
+
+// The Kiali ServiceAccount token.
+var saToken string
+
+var clientFactory kubernetes.ClientFactory
+
+func DiscoverJaeger() (url string, err error) {
+	if config.Get().ExternalServices.Jaeger.URL != "" {
+		return config.Get().ExternalServices.Jaeger.URL, nil
+	}
+	return discoverUrlService(config.Get().ExternalServices.Jaeger.ServiceNamespace, config.Get().ExternalServices.Jaeger.Service)
+}
+
+func DiscoverGrafana() (url string, err error) {
+	if config.Get().ExternalServices.Grafana.URL != "" {
+		return config.Get().ExternalServices.Grafana.URL, nil
+	}
+	return discoverUrlService(config.Get().ExternalServices.Grafana.ServiceNamespace, config.Get().ExternalServices.Grafana.Service)
+}
+
+func discoverUrlService(ns string, service string) (url string, err error) {
+
+	if saToken == "" {
+		token, err := ioutil.ReadFile("/var/run/secrets/kubernetes.io/serviceaccount/token")
+		if err != nil {
+			return "", err
+		}
+		saToken = string(token)
+	}
+
+	if clientFactory == nil {
+		userClientFactory, err := kubernetes.GetClientFactory()
+		if err != nil {
+			return "", err
+		}
+		clientFactory = userClientFactory
+	}
+
+	client, err := clientFactory.GetClient(saToken)
```

Comment:
I may be wrong, but I think if we don't provide any token (empty string) then it automatically creates client on behalf of Kiali, so you don't have to read the token from file. Is it wrong?

Reply:
I saw this code in grafana file so I thought that we need it . https://github.com/kiali/kiali/blob/23e4206276d090498cf45630f24881996c58dcf8/handlers/grafana.go